File: src/runtime/boxes/BbAlt.cs
// BbAlt.cs — ALT: alternation — try each child on α; β retries same child
// Mirrors src/runtime/boxes/bb_alt.c
//
// α: save cursor; try children[0].α ... children[n-1].α in order
//    on child_γ: remember which child matched; γ
//    on child_ω: restore cursor, try next child
// β: retry the same child that succeeded on α via child.β
//    on child_γ: γ
//    on child_ω: ω (no cross-child backtrack on β)

namespace Snobol4.Runtime.Boxes;

public sealed class BbAlt : IByrdBox
{
    private readonly IByrdBox[] _children;
    private int  _current;   // 0-based index of child that last fired γ
    private int  _savedPos;

    public BbAlt(params IByrdBox[] children)
    {
        _children = children ?? [];
    }

    public Spec Alpha(MatchState ms)
    {
        _savedPos = ms.Cursor;
        for (int i = 0; i < _children.Length; i++)
        {
            ms.Cursor = _savedPos;                   // restore for each arm
            var cr = _children[i].Alpha(ms);
            if (!cr.IsFail)
            {
                _current = i;
                return cr;                           // child_α_γ → ALT_γ
            }
            // child_α_ω → try next
        }
        ms.Cursor = _savedPos;
        return Spec.Fail;                            // ALT_ω
    }

    public Spec Beta(MatchState ms)
    {
        // β retries the same child only (mirrors bb_alt.c ALT_β)
        var cr = _children[_current].Beta(ms);
        if (!cr.IsFail) return cr;                   // child_β_γ → ALT_γ
        return Spec.Fail;                            // ALT_ω
    }
}

File: src/runtime/boxes/BbAlt.java
package snobol4.runtime.boxes;

/**
 * BbAlt.java — ALT: alternation; try each child on α; β retries same child
 * Port of bb_alt.c / bb_alt.s
 *
 * C original:
 *   ALT_α:      position=Δ; current=1;
 *               cr = children[0](α); if empty → child_α_ω; else → child_α_γ;
 *   ALT_β:      cr = children[current-1](β);
 *               if empty → ALT_ω; else → child_β_γ;
 *   child_α_γ:  result=cr;                             goto ALT_γ;
 *   child_α_ω:  current++;
 *               if current > n                         goto ALT_ω;
 *               Δ=position;
 *               cr = children[current-1](α); if empty → child_α_ω; else → child_α_γ;
 *   child_β_γ:  result=cr;                             goto ALT_γ;
 *   ALT_γ:                                             return result;
 *   ALT_ω:                                             return spec_empty;
 */
public class BbAlt extends BbBox {
    private final BbBox[] children;
    private final int     n;
    private int           current;   /* 1-based index of active child */
    private int           position;  /* saved Δ at ALT entry          */

    public BbAlt(MatchState ms, BbBox... children) {
        super(ms);
        this.children = children;
        this.n        = children.length;
    }

    @Override public Spec alpha() {
        // ALT_α
        position = ms.delta;
        current  = 1;
        return tryAlpha();
    }

    @Override public Spec beta() {
        // ALT_β: retry same child with β
        Spec cr = children[current - 1].beta();
        if (cr == null) return null;                                   // ALT_ω
        return cr;                                                     // child_β_γ → ALT_γ
    }

    /* child_α_ω loop: advance to next child */
    private Spec tryAlpha() {
        while (current <= n) {
            ms.delta = position;
            Spec cr  = children[current - 1].alpha();
            if (cr != null) return cr;                                 // child_α_γ → ALT_γ
            current++;                                                 // child_α_ω
        }
        return null;                                                   // ALT_ω
    }
}
